controller/integration/test/parallel/parallel_test.go

```diff
@@ -0,0 +1,224 @@
+	// 10 second ResyncPeriod + 2 second RateWait + 8 second for safety.
+	// The controllers should now add their finalizers and EnsureCreated should be
+	// hit once immediatly.
+	//
+	// 		EnsureCreated: 1, EnsureDeleted: 0
+	//
+	// The controllers should reconcile once in this period.
+	//
+	// 		EnsureCreated: 2, EnsureDeleted: 0
+	//
+	operation = func() error {
+		// We are more forgiving here compared to other tests, the controllers might
+		// receive events at different times. Checking the count exactly might fail
+		// if a controller is slower and the other one reconciles one more time.
+		if trA.CreateCount() < 2 {
+			return microerror.Maskf(countMismatchError, "EnsureCreated of controller A was hit %v times, want atleast %v", trA.CreateCount(), 2)
+		}
+		if trA.DeleteCount() != 0 {
+			return microerror.Maskf(countMismatchError, "EnsureDeleted of controller A was hit %v times, want %v", trB.DeleteCount(), 0)
+		}
+		if trB.CreateCount() < 2 {
+			return microerror.Maskf(countMismatchError, "EnsureCreated of controller B was hit %v times, want atleast %v", trB.CreateCount(), 2)
+		}
+		if trB.DeleteCount() != 0 {
+			return microerror.Maskf(countMismatchError, "EnsureDeleted of controller B was hit %v times, want %v", trB.DeleteCount(), 0)
+		}
+		return nil
+	}
+	err = backoff.Retry(operation, newConstantBackoff(uint64(20)))
+	if err != nil {
+		t.Fatal("expected", nil, "got", err)
+	}
+
+	// We get the object after the controllers have been started.
+	resultObj, err := testWrapperA.GetObject(objName, testNamespace)
+	if err != nil {
+		t.Fatal("expected", nil, "got", err)
+	}
+
+	resultObjAccessor, err := meta.Accessor(resultObj)
+	if err != nil {
+		t.Fatal("expected", nil, "got", err)
+	}
+
+	// We verify, that the DeletionTimestamp has not been set.
+	if resultObjAccessor.GetDeletionTimestamp() != nil {
+		t.Fatalf("DeletionTimestamp != nil, want nil")
+	}
+
+	// We verify, that our finalizer is still set.
+	// We check this individually because we are not sure in which order the
+	// finalizers were added.
+	if !containsFinalizer(resultObjAccessor.GetFinalizers(), testFinalizerA) {
+		t.Fatalf("finalizers == %v, want to contain %v", resultObjAccessor.GetFinalizers(), testFinalizerA)
+	}
+	if !containsFinalizer(resultObjAccessor.GetFinalizers(), testFinalizerB) {
+		t.Fatalf("finalizers == %v, want to contain %v", resultObjAccessor.GetFinalizers(), testFinalizerB)
+	}
+
+	// We delete the object now.
+	err = testWrapperA.DeleteObject(objName, testNamespace)
+	if err != nil {
+		t.Fatal("expected", nil, "got", err)
+	}
+
+	// We use backoff with the absolute maximum amount:
+	// 2 second RateWait + 18 second to ensure that both finalizers were removed.
+	// The controllers should now remove the finalizer and EnsureDeleted should be
+	// hit four times immediatly.
+	// See https://github.com/giantswarm/giantswarm/issues/2897
+	//
+	// 		EnsureDeleted: 4
+	//
+	operation = func() error {
+		if trA.DeleteCount() != 4 {
```

Comment:
We can check accurately here because there is no reconciliation.
We do not care about the `CreateCount` here anymore.